Add X2Condition mapping decorator and inverse condition

Iridar opened the pull request:
Closes #699

Adds an `X2Condition_MappingDecorator` class that stores an `X2Condition` and remaps its return codes according to specified remapping data.

Adds `X2Condition_ReverseCondition`, which is an extension of the previous class, preconfigured for simply reversing the return codes of the inlaid condition.

Iridar commented:
This PR should have been a draft, it has no docs yet, and I don't think it even works. We also should probably have a discussion on names of properties and methods of the `X2Condition_MappingDecorator`, as well as the name of the class itself, because personally `mapping decorator` rings completely hollow for me, and associates with map design at best.

Iridar commented:
Yeah, this doesn't work, although I don't understand why. The InnerCondition seems to always return `AA_Success` regardless of how I set it up.

robojumper commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/X2Condition_MappingDecorator.uc:
```diff
@@ -0,0 +1,88 @@
+
+event name CallMeetsCondition(XComGameState_BaseObject kTarget) 
+{
+	return RemapCondition(InnerCondition.CallMeetsCondition(kTarget));
```
```suggestion
	return RemapCondition(InnerCondition.MeetsCondition(kTarget));
```

robojumper commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/X2Condition_MappingDecorator.uc:
```diff
@@ -0,0 +1,88 @@
+
+event name CallMeetsConditionWithSource(XComGameState_BaseObject kTarget, XComGameState_BaseObject kSource) 
+{ 
+	return RemapCondition(InnerCondition.CallMeetsConditionWithSource(kTarget, kSource));
```
```suggestion
	return RemapCondition(InnerCondition.MeetsConditionWithSource(kTarget, kSource));
```

robojumper commented on X2WOTCCommunityHighlander/Src/XComGame/Classes/X2Condition_MappingDecorator.uc:
```diff
@@ -0,0 +1,88 @@
+
+event name CallAbilityMeetsCondition(XComGameState_Ability kAbility, XComGameState_BaseObject kTarget) 
+{
+	return RemapCondition(InnerCondition.CallAbilityMeetsCondition(kAbility, kTarget));
```
```suggestion
	return RemapCondition(InnerCondition.AbilityMeetsCondition(kAbility, kTarget));
```

robojumper reviewed (changes requested):
I believe the problem here is that you're unconditionally using the UnrealScript implementation of `InnerCondition` even if it's a native condition. For reference, this is in `X2Condition`:

```unrealscript
native function name MeetsCondition(XComGameState_BaseObject kTarget);
native function name MeetsConditionWithSource(XComGameState_BaseObject kTarget, XComGameState_BaseObject kSource);
native function name AbilityMeetsCondition(XComGameState_Ability kAbility, XComGameState_BaseObject kTarget);

/*
 * These should ONLY be called from the native functions above.
 */
event name CallMeetsCondition(XComGameState_BaseObject kTarget) { return 'AA_Success'; }
event name CallMeetsConditionWithSource(XComGameState_BaseObject kTarget, XComGameState_BaseObject kSource) { return 'AA_Success'; }
event name CallAbilityMeetsCondition(XComGameState_Ability kAbility, XComGameState_BaseObject kTarget) { return 'AA_Success'; }
```

The general idea is that the native (or [denativized](https://github.com/X2CommunityCore/X2WOTCCommunityHighlander/pull/918)) `XCGS_Ability:GatherAbilityTargets` calls the native `MeetsCondition`/`MeetsConditionWithSource`/`AbilityMeetsCondition`. If the `X2Condition` subclass in question is implemented with native code, this allows evaluating the condition without ever leaving C++ and calling into UnrealScript.

Only if the `X2Condition` subclass doesn't override a function, the base implementation in `X2Condition` calls the respective `Call*` variant of the function, which allows UnrealScript code to do things.

Result: You should make these three changes :)

Iridar commented:
Oh snap. Looks like you already guarded me against this error for [this wiki article](https://www.reddit.com/r/xcom2mods/wiki/index/conditions#wiki_using_conditions_inside_conditions). Thanks!

Iridar commented:
Looks like this still doesn't work for a typical expected use case. 

I made a copy of standard shot as a test ability, removed all other conditions from it, and then did this:
```
	ReverseCondition = new class'X2Condition_ReverseCondition';

	UnitProperty = new class'X2Condition_UnitProperty';
	UnitProperty.ExcludeLargeUnits = true;
	UnitProperty.ExcludeDead=false;
	UnitProperty.ExcludeFriendlyToSource=false;
	UnitProperty.ExcludeCosmetic=false;
	UnitProperty.ExcludeInStasis=false;

	ReverseCondition.InnerCondition = UnitProperty;
	ReverseCondition.bUseCanEverBeValid = false;

	Template.AbilityTargetConditions.AddItem(ReverseCondition);
```
Essentially, I'm adding a unit property condition that - if used normally - would allow you to target only regular-sized units, e.g. can target soldiers and troopers, but not large units, like Sectopods. This does work as expected when I add the UnitProperty condition directly, without reversing it.

So when reversed you would expect to be able to target *only* Sectopods. However, in testing the ability cannot target *any* units. The logs from it are:
```
[0043.26] IRITEST: Remapping Condition: X2Condition_UnitProperty original return code: AA_UnitIsTooLarge
[0043.26] IRITEST: Did not find remapped code, defaulting to: AA_Success
[0043.26] IRITEST: Remapping Condition: X2Condition_UnitProperty original return code: AA_Success
[0043.26] IRITEST: Found remapped code: 0 : AA_AbilityUnavailable
[0043.26] IRITEST: Remapping Condition: X2Condition_UnitProperty original return code: AA_Success
[0043.26] IRITEST: Found remapped code: 0 : AA_AbilityUnavailable
[0043.26] IRITEST: Remapping Condition: X2Condition_UnitProperty original return code: AA_Success
[0043.26] IRITEST: Found remapped code: 0 : AA_AbilityUnavailable
[0043.26] IRITEST: Remapping Condition: X2Condition_UnitProperty original return code: AA_Success
[0043.26] IRITEST: Found remapped code: 0 : AA_AbilityUnavailable
[0043.26] IRITEST: Remapping Condition: X2Condition_UnitProperty original return code: AA_Success
[0043.26] IRITEST: Found remapped code: 0 : AA_AbilityUnavailable
[0043.26] IRITEST: Remapping Condition: X2Condition_UnitProperty original return code: AA_Success
[0043.26] IRITEST: Found remapped code: 0 : AA_AbilityUnavailable
```
As we can see, the remapped condition does return `AA_Success` once, presumably when the condition is checked against the Sectopod I have dropped in addition to a Trooper, an allied Muton and an allied SPARK with BIT and a Templar.

My only guess here is what's happening is that the property condition returns `AA_UnitIsTooLarge` in `UnitProperty.MeetsCondition()`, but returns `AA_Success` in `UnitProperty.MeetsConditionWithSource()`, which then gets remapped to `AA_AbilityUnavailable`.

So it seems we have a design issue on our hands when dealing with conditions that implement several checks at the same time.

We either have to remap them separately or opt out of remapping them. Both would complicate things for the user. 

This issue was originally supposed to make reversing conditions easier, but if this thing is gonna require a doctorate in programming to use properly, it might not be worth the effort. Further discussion on the topic is welcome, because I'm not sure what to do at this point.

robojumper commented:
@Iridar Oof, that's a pretty unfortunate oversight indeed. On a more abstract level, an ability is available for a unit and target if and only if all operators of all conditions return `AA_Success`. This is conceptually a logical conjunction of the condition results, each of which is a conjunction of the return values of `MeetsCondition`, `MeetsConditionWithSource`, `AbilityMeetsCondition`. We'd like to invert one condition, so that's basically:

`!(Condition.MeetsCondition(...) && Condition.MeetsConditionWithSource(...) && Condition.AbilityMeetsCondition(...))`

Since the overall condition checking happens in native code, we have no way to invert the result of the conjunction. We could break the term down further:

`(!Condition.MeetsCondition(...) || !Condition.MeetsConditionWithSource(...) || !Condition.AbilityMeetsCondition(...))`

But that runs afoul of the exact same issue -- we can't simply use a disjunction because it's native code. The only way is -- as you said -- "remap [the operators] separately".

> So it seems we have a design issue on our hands when dealing with conditions that implement several checks at the same time.

I'm not sure what you mean. Every condition always implements the three operators and returns `AA_Success` unless it has a reason not to. A "is not large" check turns `Condition.MeetsCondition(...) && Condition.MeetsConditionWithSource(...) && Condition.AbilityMeetsCondition(...)` into `Unit.NotLarge() && true && true` because the operators yielding an unconditional `true` are fundamentally a "don't care" that our operators then turn into a `false`.

The "don't care" thing is a new insight for me. I guess the only way to implement a system that supports these decorators is to introduce three-valued logic with a "AA_DontCare" or something.
The "remapping separately" comes close. By remapping operators selectively, we essentially turn the unconditional `AA_Success` into a "don't care", but as you said, this requires insight into the internal workings of the decorated condition (most of which are native). At this point I think this is pretty hard to get right and custom conditions should be preferred. Recommending closing the PR and issue as infeasible.

Iridar commented:
Closing per robojumper's recommendation.
